cri/v1alpha2/service.go

```diff
@@ -0,0 +1,76 @@
+package v1alpha2
+
+import (
+	"context"
+	"path"
+
+	runtime "github.com/alibaba/pouch/cri/apis/v1alpha2"
+	"github.com/alibaba/pouch/cri/metrics"
+	"github.com/alibaba/pouch/daemon/config"
+	"github.com/alibaba/pouch/pkg/grpc/interceptor"
+	"github.com/alibaba/pouch/pkg/netutils"
+
+	"github.com/grpc-ecosystem/go-grpc-middleware"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+)
+
+// Service serves the kubelet runtime grpc api which will be consumed by kubelet.
+type Service struct {
+	config *config.Config
+	server *grpc.Server
+}
+
+// NewService creates a brand new cri service.
+func NewService(cfg *config.Config, criMgr CriMgr) (*Service, error) {
+	logEntry := logrus.NewEntry(logrus.StandardLogger())
+
+	s := &Service{
+		config: cfg,
+		server: grpc.NewServer(
+			grpc.StreamInterceptor(metrics.GRPCMetrics.StreamServerInterceptor()),
+			grpc_middleware.WithUnaryServerChain(
+				metrics.GRPCMetrics.UnaryServerInterceptor(),
+				interceptor.PayloadUnaryServerInterceptor(logEntry, criLogLevelDecider),
+			),
+		),
+	}
+
+	runtime.RegisterRuntimeServiceServer(s.server, criMgr)
+	runtime.RegisterImageServiceServer(s.server, criMgr)
+	runtime.RegisterVolumeServiceServer(s.server, criMgr)
+
+	// EnableHandlingTimeHistogram turns on recording of handling time
+	// of RPCs. Histogram metrics can be very expensive for Prometheus
+	// to retain and query.
+	metrics.GRPCMetrics.EnableHandlingTimeHistogram()
+	// Initialize all metrics.
+	metrics.GRPCMetrics.InitializeMetrics(s.server)
+
+	return s, nil
+}
+
+// Serve starts grpc server.
+func (s *Service) Serve() error {
+	l, err := netutils.GetListener(s.config.CriConfig.Listen, nil)
+	if err != nil {
+		return err
+	}
+
+	return s.server.Serve(l)
+}
+
+func criLogLevelDecider(ctx context.Context, fullMethodName string, servingObject interface{}) logrus.Level {
+	// extract methodName from fullMethodName
+	// eg. extract 'StartContainer' from '/runtime.v1alpha2.RuntimeService/StartContainer'
+	methodName := path.Base(fullMethodName)
+
+	// method->logLevel map
+	switch methodName {
+	case "Version", "PodSandboxStatus", "ListPodSandbox", "ListContainers", "ContainerStatus",
```

Comment:
I think one item in one line will be good to review. If we add one new one or remove old one, it will be only one line diff. WDYT? it is not big deal, and just open to discuss.

Reply:
agree with you. One line per entry is vcs friendly.